test: fix p2p_leak_tx.py

mzumsande opened the pull request:
This fixes two issues with `p2p_leak_tx.py`:

1.) #33090: As far as I can see, this is just the randomness of `NextInvToInbounds`/ `rand_exp_duration`, which has a probability of `e^-(60s/5s) = 6.14×10^−6` to result in a period > 60s (our waiting time), so that the test would fail every 160k runs... Doubling the timeout should be sufficient to lower the probability drastically.

2.) The subtest `test_notfound_on_unannounced_tx` has some (w)txid confusion: we send a `MSG_TX`-type getdata with a `wtxid` in it, which necessarily always results in a NOTFOUND. Fixed this, and change the subtest to be more deterministic based on `mocktime`.

mzumsande committed:
test: fix (w)txid confusion in p2p_leak_tx.py

Before, we'd send a MSG_TX with a wtxid in it, which
would always result in a notfound answer

mzumsande committed:
test: increase timeout in p2p_leak_tx.py

With a low but not negligible probability in the order
of 10^-6 the exponential timer NextInvToInBounds can lead
to an interval >60s, making the test fail.
Also uses mocktime to speed up the test and fixes a
non-matching on_inv override.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

fanquake commented:
MSAN failure (https://cirrus-ci.com/task/6435705422872576?logs=ci#L1863) is #33114.

jonatack commented:
Concept ACK

stratospher reviewed (commented):
Nice detective work! Just 1 question, will ACK after that.

1. I tried some modifications of the test locally but wasn't able to enter the "else branch" where "tx was already announced to us". Were you able to test it?
2. Could also change MAX_REPEATS to a lower number. I'm getting `bad-txns-inputs-missingorspent` at run 47 when I remove the break in the loop. So don't think the test would work for full 100 runs anyways.

> Not sure if this is essential for the test, but we could add a random delay of 2-3s to trigger both paths on a regular basis.

Hmm I'm fine either ways. On one hand - point of the test is to make sure that node responds with NOTFOUND for unannounced transactions, on the other hand - silent errors like the one fixed in this PR can slip through undetected.

mzumsande commented:
> I tried some modifications of the test locally but wasn't able to enter the "else branch" where "tx was already announced to us". Were you able to test it?

No, but with a `time.sleep(3)` before `inbound_peer.send_and_ping(want_tx)` it would go into that branch regularly. because the `NextInvToInbounds` will often generate a random interval `<3s` - that's why I mentioned it above.

In general I was a bit unsure about the main point of the subtest (maybe @maflcko you can help?). Is the goal to have both branches executed regularly, or are we just interested in the `notfound` branch and the rest is there mostly to avoid intermittent failure?
By the way, the failure from 1) could also be easily avoided by setting `self.noban_tx_relay = True`, but that seemed to be at odds with the general intention of the test.

> Could also change MAX_REPEATS to a lower number

True, depending on whether we put a sleep in or not.

naiyoma reviewed (commented):
Concept ACK 

Increasing  timeout to 120 reduces probability to 
`P(120s) = e^(-120/5) = e^(-24) ≈ 3.78×10^-11 (~1 in 26.5B runs)`

naiyoma commented:
> but with a `time.sleep(3)` before `inbound_peer.send_and_ping(want_tx)` it would go into that branch regularly.

`time.sleep(8)` right before `wtxid = self.miniwallet.send_self_transfer(from_node=self.gen_node)["wtxid"]` also works

enirox001 reviewed (commented):
Concept ACK ab43e0b on fixing the test flakiness.

I took a shot at making `test_notfound_on_unannounced_tx()` more deterministic using mocktime. Feel free to use it if helpful:

```diff
diff --git a/test/functional/p2p_leak_tx.py b/test/functional/p2p_leak_tx.py
index 3fd324a029..f50993e932 100755
--- a/test/functional/p2p_leak_tx.py
+++ b/test/functional/p2p_leak_tx.py
@@ -11,6 +11,7 @@ from test_framework.util import (
     assert_equal,
 )
 from test_framework.wallet import MiniWallet
+import time
 
 
 class P2PNode(P2PDataStore):
@@ -79,27 +80,33 @@ class P2PLeakTxTest(BitcoinTestFramework):
         self.gen_node.disconnect_p2ps()
         inbound_peer = self.gen_node.add_p2p_connection(P2PNode())  # An "attacking" inbound peer
 
-        MAX_REPEATS = 100
-        self.log.info("Running test up to {} times.".format(MAX_REPEATS))
-        for i in range(MAX_REPEATS):
-            self.log.info('Run repeat {}'.format(i + 1))
-            wtxid = self.miniwallet.send_self_transfer(from_node=self.gen_node)["wtxid"]
-
-            want_tx = msg_getdata()
-            want_tx.inv.append(CInv(t=MSG_WTX, h=int(wtxid, 16)))
-            with p2p_lock:
-                inbound_peer.last_message.pop('notfound', None)
-            inbound_peer.send_and_ping(want_tx)
-            if inbound_peer.last_message.get('notfound'):
-                self.log.debug('tx {} was not yet announced to us.'.format(wtxid))
-                self.log.debug("node has responded with a notfound message. End test.")
-                assert_equal(inbound_peer.last_message['notfound'].vec[0].hash, int(wtxid, 16))
+        MOCK_DELTA = 120
+        MAX_REPEATS = 10
+        try:
+            self.gen_node.setmocktime(int(time.time()) + MOCK_DELTA)
+            for i in range(MAX_REPEATS):
+                self.log.info('Run repeat {}'.format(i + 1))
+                wtxid = self.miniwallet.send_self_transfer(from_node=self.gen_node)["wtxid"]
+                want_tx = msg_getdata()
+                want_tx.inv.append(CInv(t=MSG_WTX, h=int(wtxid, 16)))
                 with p2p_lock:
-                    inbound_peer.last_message.pop('notfound')
-                break
+                    inbound_peer.last_message.pop('notfound', None)
+                    inbound_peer.last_message.pop('tx', None)
+                inbound_peer.send_and_ping(want_tx)
+                inbound_peer.wait_until(lambda: inbound_peer.last_message.get('notfound') or inbound_peer.last_message.get('tx'), timeout=10)
+                if inbound_peer.last_message.get('notfound'):
+                    self.log.debug('tx {} was not yet announced to us.'.format(wtxid))
+                    assert_equal(inbound_peer.last_message['notfound'].vec[0].hash, int(wtxid, 16))
+                    with p2p_lock:
+                        inbound_peer.last_message.pop('notfound')
+                    break
+                else:
+                    self.log.debug('tx {} was already announced to us. Try test again.'.format(wtxid))
+                    assert int(wtxid, 16) in [inv.hash for inv in inbound_peer.last_message['inv'].inv]
             else:
-                self.log.debug('tx {} was already announced to us. Try test again.'.format(wtxid))
-                assert int(wtxid, 16) in [inv.hash for inv in inbound_peer.last_message['inv'].inv]
+                raise AssertionError("Test failed after {} attempts - transactions kept being announced".format(MAX_REPEATS))
+        finally:
+            self.gen_node.setmocktime(0)
 
 
 if __name__ == '__main__':
```

This uses mocktime to control announcement timing more deterministically while keeping a small retry loop for robustness.

davidgumberg committed:
test: make notfound_on_unannounced more reliable

By using mocktime, we will always hit both the notfound
branch and the tx sent branch.
The previous version didn't achieve that due to timing
issues.

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

mzumsande reviewed (commented):
I have now pushed an update - I went with the approach of @davidgumberg because it's simpler - when we use mocktime, `MAX_REPEATS` should no longer be necessary.

I made small changes to the suggestion above:
- used current time with `setmocktime`, going back in time seems brittle.
-  I didn't understand the comment "Checking that the inv contains the wtxid would have a 1 in 10^6 chance of spuriously failing (...)", so I didn't take that part.
 If `NextInvToInbounds` doesn't fire, we should just get no inv and the `wait_until` should fail (I also used 120s timeout here to make that less likely), but if we receive an inv, shouldn't it always contain the wtxid?

davidgumberg commented:
> If `NextInvToInbounds` doesn't fire, we should just get no inv and the `wait_until` should fail (I also used 120s timeout here to make that less likely), but if we receive an inv, shouldn't it always contain the wtxid?

My understanding of [`Peer::TxRelay::m_tx_inventory_known_filter`](https://github.com/bitcoin/bitcoin/blob/e04cb9c1bdf2199127cc8cf9c87f25e46b8cac7b/src/net_processing.cpp#L297-L300) is that it will [occasionally indicate](https://github.com/bitcoin/bitcoin/blob/e04cb9c1bdf2199127cc8cf9c87f25e46b8cac7b/src/common/bloom.h#L91-L92) that a transaction is known to a peer when it isn't, and then we won't include it in an INV:

https://github.com/bitcoin/bitcoin/blob/e04cb9c1bdf2199127cc8cf9c87f25e46b8cac7b/src/net_processing.cpp#L5784-L5786

But I guess my comment was mistaken, since in this particular test since there is only one transaction to announce and that would result in the peer not sending any INV message at all:

https://github.com/bitcoin/bitcoin/blob/e04cb9c1bdf2199127cc8cf9c87f25e46b8cac7b/src/net_processing.cpp#L5807-L5808

But there are other tests that wait for INV's with exact contents, e.g.: https://github.com/bitcoin/bitcoin/blob/e04cb9c1bdf2199127cc8cf9c87f25e46b8cac7b/test/functional/p2p_segwit.py#L1263-L1267

and I can't find any reports about them spuriously failing, so maybe I'm misunderstanding something or this failure case is sufficiently rare.

mzumsande commented:
> My understanding of [`Peer::TxRelay::m_tx_inventory_known_filter`](https://github.com/bitcoin/bitcoin/blob/e04cb9c1bdf2199127cc8cf9c87f25e46b8cac7b/src/net_processing.cpp#L297-L300) is that it will [occasionally indicate](https://github.com/bitcoin/bitcoin/blob/e04cb9c1bdf2199127cc8cf9c87f25e46b8cac7b/src/common/bloom.h#L91-L92) that a transaction is known to a peer when it isn't, and then we won't include it in an INV:

Good point, I didn't think of that possibility.

> Maybe this is also a suspect for the spurious failure in https://github.com/bitcoin/bitcoin/issues/33090?

I don't think we can determine what happened in the failed run, but `10^-6` is 6x smaller than the probability of `NextInvToInbounds > 60s`, so it still seems ok to bump the timeout.

davidgumberg reviewed (approved):
ACK e14825ace2a8c6abcf11fdfb

- (https://github.com/bitcoin/bitcoin/pull/33121/commits/d73952df9e755f1349ff4a19c47ffc92efa3a2b1) Mitigates the likeliest cause of the spurious failure #33090 in `p2p_leak_tx.py` by increasing the timeout we'll wait for an INV from 60s to 120s.
- (https://github.com/bitcoin/bitcoin/pull/33121/commits/3005fb2ad39aec6fdb52dc6f406672a9515934ca) `test_notfound_on_unannounced_tx` previously was not really testing anything, as written in the PR description: 
    > we send a MSG_TX-type getdata with a wtxid in it, which necessarily always results in a NOTFOUND
- (https://github.com/bitcoin/bitcoin/pull/33121/commits/e14825ace2a8c6abcf11fdfb5057e969cd64b146) Instead of probabilistically testing one of two branches, this uses `setmocktime` to explicitly test that sending a `getdata` for a transaction that has not yet been announced gets a `NOTFOUND` response, and that sending a `getdata` for a transaction that *has* been announced, we receive a `TX` message with the transaction.

davidgumberg reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/commit/a52c148ece2296f22e2fe1b7e9584d49e23f03f2

vasild commented on test/functional/p2p_leak_tx.py:
```diff
@@ -47,7 +47,7 @@ def test_tx_in_block(self):
         self.gen_node.setmocktime(0)
 
-        inbound_peer.wait_until(lambda: "inv" in inbound_peer.last_message and inbound_peer.last_message.get("inv").inv[0].hash == int(wtxid, 16))
+        inbound_peer.wait_until(lambda: "inv" in inbound_peer.last_message and inbound_peer.last_message.get("inv").inv[0].hash == int(wtxid, 16), timeout=120)
```
Here is a change against `master` that could replace the first commit in this PR. It solves the same problem - when `NextInvToInbounds()` decides to wait more than the test timeout. In addition, by using mocktime, the diff below eliminates the waits in a few places of the test due to the `NextInvToInbounds()` delays. As a result the test passes in about 1 second instead of the current random few seconds to few tens of seconds.

```diff
--- i/test/functional/p2p_leak_tx.py
+++ w/test/functional/p2p_leak_tx.py
@@ -24,31 +24,34 @@ class P2PLeakTxTest(BitcoinTestFramework):
         self.num_nodes = 1
 
     def run_test(self):
         self.gen_node = self.nodes[0]  # The block and tx generating node
         self.miniwallet = MiniWallet(self.gen_node)
 
+        self.mocktime = int(time.time())
+
         self.test_tx_in_block()
         self.test_notfound_on_replaced_tx()
         self.test_notfound_on_unannounced_tx()
 
     def test_tx_in_block(self):
         self.log.info("Check that a transaction in the last block is uploaded (beneficial for compact block relay)")
+        self.gen_node.setmocktime(self.mocktime)
         inbound_peer = self.gen_node.add_p2p_connection(P2PNode())
 
         self.log.debug("Generate transaction and block")
         inbound_peer.last_message.pop("inv", None)
 
-        self.gen_node.setmocktime(int(time.time())) # pause time based activities
         wtxid = self.miniwallet.send_self_transfer(from_node=self.gen_node)["wtxid"]
         rawmp = self.gen_node.getrawmempool(False, True)
         pi = self.gen_node.getpeerinfo()[0]
         assert_equal(rawmp["mempool_sequence"], 2) # our tx cause mempool activity
         assert_equal(pi["last_inv_sequence"], 1) # that is after the last inv
         assert_equal(pi["inv_to_send"], 1) # and our tx has been queued
-        self.gen_node.setmocktime(0)
+        self.mocktime += 120
+        self.gen_node.setmocktime(self.mocktime)
 
         inbound_peer.wait_until(lambda: "inv" in inbound_peer.last_message and inbound_peer.last_message.get("inv").inv[0].hash == int(wtxid, 16))
 
         rawmp = self.gen_node.getrawmempool(False, True)
         pi = self.gen_node.getpeerinfo()[0]
         assert_equal(rawmp["mempool_sequence"], 2) # no mempool update
@@ -62,21 +65,26 @@ class P2PLeakTxTest(BitcoinTestFramework):
         inbound_peer.last_message.pop("tx", None)
         inbound_peer.send_and_ping(want_tx)
         assert_equal(inbound_peer.last_message.get("tx").tx.wtxid_hex, wtxid)
 
     def test_notfound_on_replaced_tx(self):
         self.gen_node.disconnect_p2ps()
+        self.gen_node.setmocktime(self.mocktime)
         inbound_peer = self.gen_node.add_p2p_connection(P2PTxInvStore())
 
         self.log.info("Transaction tx_a is broadcast")
         tx_a = self.miniwallet.send_self_transfer(from_node=self.gen_node)
-        inbound_peer.wait_for_broadcast(txns=[tx_a["wtxid"]])
+        self.mocktime += 120
+        self.gen_node.setmocktime(self.mocktime)
+        inbound_peer.wait_until(lambda: "inv" in inbound_peer.last_message and inbound_peer.last_message.get("inv").inv[0].hash == int(tx_a['wtxid'], 16))
 
         tx_b = tx_a["tx"]
         tx_b.vout[0].nValue -= 9000
         self.gen_node.sendrawtransaction(tx_b.serialize().hex())
+        self.mocktime += 120
+        self.gen_node.setmocktime(self.mocktime)
         inbound_peer.wait_until(lambda: "tx" in inbound_peer.last_message and inbound_peer.last_message.get("tx").tx.wtxid_hex == tx_b.wtxid_hex)
 
         self.log.info("Re-request of tx_a after replacement is answered with notfound")
         req_vec = [
             CInv(t=MSG_TX, h=int(tx_a["txid"], 16)),
             CInv(t=MSG_WTX, h=int(tx_a["wtxid"], 16)),
```

mzumsande replied:
done with the latest push. I also adjusted the third commit with the new test in a similar way.

vasild reviewed (approved):
ACK a52c148ece2296f22e2fe1b7e9584d49e23f03f2

The changes are an improvement. It seems some of that maybe can be done better, see https://github.com/bitcoin/bitcoin/pull/33121#discussion_r2381936863. I do not see that as a blocker.

mzumsande commented:
[a52c148](https://github.com/bitcoin/bitcoin/commit/a52c148ece2296f22e2fe1b7e9584d49e23f03f2) to [14ae71f](https://github.com/bitcoin/bitcoin/commit/14ae71f323dd011c6d51470ea15cf00750970f65):
- Used mocktime everywhere to speed up the test.
- corrected `on_inv` override variable name in 1st commit (my IDE complained about that)

vasild reviewed (approved):
ACK 14ae71f323dd011c6d51470ea15cf00750970f65
